File: src/voice_kb/geometry.py
"""Pure placement math for the dictation window.

No ``xrandr``, no subprocess: the caller (the imperative shell) is
responsible for discovering real output geometry and the pointer position,
and hands them in as plain :class:`Rect` values. That's what makes this
testable without an X server.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class Rect:
    """An axis-aligned pixel rectangle: X11 output geometry, a window, or
    the pointer as a 1x1 anchor.

    ``width``/``height`` rather than a second corner point, matching how
    ``xrandr`` reports geometry -- a rect with zero or negative area cannot
    represent a real monitor or window, so it's rejected here rather than
    producing nonsense areas downstream.
    """

    x: int
    y: int
    width: int
    height: int

    def __post_init__(self) -> None:
        if self.width <= 0 or self.height <= 0:
            raise ValueError(f"Rect must have positive area: {self.width}x{self.height}")

    @property
    def right(self) -> int:
        return self.x + self.width

    @property
    def bottom(self) -> int:
        return self.y + self.height

    def intersection_area(self, other: Rect) -> int:
        overlap_w = max(0, min(self.right, other.right) - max(self.x, other.x))
        overlap_h = max(0, min(self.bottom, other.bottom) - max(self.y, other.y))
        return overlap_w * overlap_h


@dataclass(frozen=True, slots=True)
class Output:
    """One monitor, as reported by ``xrandr``."""

    name: str
    rect: Rect
    primary: bool = False


def pick_output(outputs: Sequence[Output], window: Rect) -> Output:
    """Choose the output that holds ``window`` -- for the dictation window,
    a 1x1 rect at the mouse pointer.

    The output with the largest intersection with ``window`` wins. If the
    window doesn't overlap any output at all (off-screen, or the caller
    passed stale geometry), fall back to the primary output, or the first
    one if none is marked primary.
    """
    if not outputs:
        raise ValueError("pick_output requires at least one output")
    best = max(outputs, key=lambda o: o.rect.intersection_area(window))
    if best.rect.intersection_area(window) > 0:
        return best
    return next((o for o in outputs if o.primary), outputs[0])


def dictation_rect(output: Rect, anchor: tuple[int, int] | None, fraction: float) -> Rect:
    """Where the dictation window goes on ``output``.

    ``fraction`` scales each axis, so the default of 0.5 gives a window
    covering a quarter of the screen's area. ``anchor`` is the mouse pointer,
    used as the window's *top-left* corner: the window appears next to what
    the user is reading rather than in a fixed corner they have to look away
    to find.

    Clamped into ``output``: a pointer near the right or bottom edge would
    otherwise anchor a window that hangs off the screen. Clamping means a
    pointer in the bottom-right corner lands the window flush in the
    bottom-right quarter, which is also the placement used when ``anchor`` is
    ``None`` because the pointer could not be read.
    """
    width = max(1, int(output.width * fraction))
    height = max(1, int(output.height * fraction))
    if anchor is None:
        x, y = output.right - width, output.bottom - height
    else:
        x, y = anchor
    x = max(output.x, min(x, output.right - width))
    y = max(output.y, min(y, output.bottom - height))
    return Rect(x=x, y=y, width=width, height=height)

File: src/geometry.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Rect {
    pub x: i32,
    pub y: i32,
    pub width: u32,
    pub height: u32,
}
impl Rect {
    pub fn intersection(self, other: Self) -> i64 {
        let w = (i64::from(self.x) + i64::from(self.width))
            .min(i64::from(other.x) + i64::from(other.width))
            - i64::from(self.x.max(other.x));
        let h = (i64::from(self.y) + i64::from(self.height))
            .min(i64::from(other.y) + i64::from(other.height))
            - i64::from(self.y.max(other.y));
        w.max(0) * h.max(0)
    }
}
#[derive(Debug, Clone)]
pub struct Output {
    pub rect: Rect,
    pub primary: bool,
}
pub fn pick_output(outputs: &[Output], anchor: Rect) -> Option<Rect> {
    let best = outputs
        .iter()
        .enumerate()
        .max_by_key(|(i, o)| (o.rect.intersection(anchor), std::cmp::Reverse(*i)))?
        .1;
    Some(
        if best.rect.intersection(anchor) > 0 {
            best
        } else {
            outputs.iter().find(|o| o.primary).unwrap_or(&outputs[0])
        }
        .rect,
    )
}
pub fn placement(output: Rect, anchor: Option<(i32, i32)>, fraction: f64) -> Rect {
    let width = (f64::from(output.width) * fraction) as u32;
    let height = (f64::from(output.height) * fraction) as u32;
    let width = width.max(1).min(output.width);
    let height = height.max(1).min(output.height);
    let right = output.x + (output.width - width) as i32;
    let bottom = output.y + (output.height - height) as i32;
    let (x, y) = anchor.unwrap_or((right, bottom));
    Rect {
        x: x.clamp(output.x, right),
        y: y.clamp(output.y, bottom),
        width,
        height,
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn negative_monitor_and_clamping() {
        let r = Rect {
            x: -1920,
            y: 0,
            width: 1920,
            height: 1080,
        };
        assert_eq!(
            placement(r, Some((-1, 1079)), 0.5),
            Rect {
                x: -960,
                y: 540,
                width: 960,
                height: 540
            }
        );
        assert_eq!(
            pick_output(
                &[Output {
                    rect: r,
                    primary: true
                }],
                Rect {
                    x: 9999,
                    y: 9999,
                    width: 1,
                    height: 1
                }
            ),
            Some(r)
        );
        assert_eq!(pick_output(&[], r), None);
    }
}
